File: benches/bench-bits/bits.rs
use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::time::Instant;

use anyhow::{bail, Result};

fn main() -> Result<()> {
    let args: Vec<_> = env::args().collect();
    if args.len() != 3 {
        bail!("Usage: bits <data file> <algo>");
    }
    let filename = &args[1];
    let algo = &args[2];
    let file = File::open(filename)?;

    let start_load = Instant::now();
    let mut strings = vec![];
    let mut partitions = vec![];
    for line in BufReader::new(file).lines() {
        let line = line?;
        let parts: Vec<_> = line.split(" ").collect();
        if parts.len() != 2 {
            bail!("Data line must have two fields");
        }
        strings.push(parts[0].to_string());
        partitions.push(parts[1].parse::<u32>()?);
    }
    let load_time = Instant::now() - start_load;
    println!("Parsed data in {:?}", load_time);

    let start_compute = Instant::now();
    let mut our_partitions = vec![0; partitions.len()];
    match algo.as_str() {
        "array" => array_partition(&strings, &mut our_partitions),
        "bitmask" => bitmask_partition(&strings, &mut our_partitions),
        "array_noclear" => array_noclear_partition(&strings, &mut our_partitions),
        _ => bail!("Unknown algo {}", algo),
    }
    let compute_time = Instant::now() - start_compute;

    for i in 0..partitions.len() {
        if partitions[i] != our_partitions[i] {
            bail!(
                "Different partitions for {} (line {}): {} vs {}",
                strings[i],
                i,
                partitions[i],
                our_partitions[i]
            );
        }
    }

    println!("Computed using algo {} in {:?}", algo, compute_time);

    Ok(())
}

fn array_partition(strings: &[String], our_partitions: &mut [u32]) {
    let mut mask = [false; 256];
    for (s, out) in strings.iter().zip(our_partitions) {
        let mut ret = 1u32;
        mask.fill(false);
        for &b in s.as_bytes() {
            if mask[b as usize] {
                mask.fill(false);
                ret += 1;
            }
            mask[b as usize] = true
        }
        *out = ret
    }
}

fn bitmask_partition(strings: &[String], our_partitions: &mut [u32]) {
    for (s, out) in strings.iter().zip(our_partitions) {
        let mut ret = 1u32;
        let mut mask = 0u32;
        for &b in s.as_bytes() {
            let m = 1u32 << (b - b'A');
            if mask & m != 0 {
                mask = 0;
                ret += 1;
            }
            mask |= m;
        }
        *out = ret
    }
}

fn array_noclear_partition(strings: &[String], our_partitions: &mut [u32]) {
    let mut prev = [0u32; 256];
    for (s, out) in strings.iter().zip(our_partitions) {
        prev.fill(0);
        let mut ret = 1u32;
        for &b in s.as_bytes() {
            if prev[b as usize] == ret {
                ret += 1;
            }
            prev[b as usize] = ret
        }
        *out = ret
    }
}

File: benches/bench-bits/bits.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
	"time"
)

func main() {
	if len(os.Args) != 3 {
		log.Fatalf("Usage: %s <data file> <algo>\n", os.Args[0])
	}
	filename := os.Args[1]
	algo := os.Args[2]

	f, err := os.Open(filename)
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()
	startLoad := time.Now()
	scanner := bufio.NewScanner(f)
	var strs []string
	var partitions []int
	for scanner.Scan() {
		v := strings.Split(scanner.Text(), " ")
		if len(v) != 2 {
			continue
		}
		strs = append(strs, v[0])
		p, err := strconv.Atoi(v[1])
		if err != nil {
			log.Fatal(err)
		}
		partitions = append(partitions, p)
	}
	fmt.Printf("Parsed data in %v\n", time.Since(startLoad))

	ourPartitions := make([]int, len(partitions))
	startCompute := time.Now()
	switch algo {
	case "array":
		arrayPartition(strs, ourPartitions)
	case "bitmask":
		bitmaskPartition(strs, ourPartitions)
	case "array_noclear":
		arrayNoclearPartition(strs, ourPartitions)
	}
	computeTime := time.Since(startCompute)
	for i, v := range ourPartitions {
		if partitions[i] != v {
			log.Fatalf("Different partitions for %s (line %d): %d vs %d\n", strs[i], i, partitions[i], v)
		}
	}
	log.Printf("Computed using algo %s in %v\n", algo, computeTime)
}

func arrayPartition(strs []string, partitions []int) {
	var mask [256]bool
	for i, s := range strs {
		for j := 0; j < 256; j++ {
			mask[j] = false
		}
		ret := 1
		for _, b := range s {
			if mask[b] {
				for j := 0; j < 256; j++ {
					mask[j] = false
				}
				ret++
			}
			mask[b] = true
		}
		partitions[i] = ret
	}
}

func bitmaskPartition(strs []string, partitions []int) {
	for i, s := range strs {
		mask := 0
		ret := 1
		for _, b := range s {
			m := 1 << (b - 'A')
			if mask&m != 0 {
				mask = m
				ret++
			} else {
				mask |= m
			}
		}
		partitions[i] = ret
	}
}

func arrayNoclearPartition(strs []string, partitions []int) {
	var prev [256]int32
	for i, s := range strs {
		for j := 0; j < 256; j++ {
			prev[j] = 0
		}
		ret := 1
		for _, b := range s {
			if prev[b] == int32(ret) {
				ret++
			}
			prev[b] = int32(ret)
		}
		partitions[i] = ret
	}
}
